extension/manifest/_manifest.py

```diff
@@ -0,0 +1,206 @@
+            raise ValueError(
+                f"Invalid magic. Expected {_ManifestLayout.EXPECTED_MAGIC}. Got {self.magic}"
+            )
+        if self.length < _ManifestLayout.EXPECTED_MIN_LENGTH:
+            raise ValueError(
+                f"Invalid length. Expected at least {_ManifestLayout.EXPECTED_MIN_LENGTH}. Got {self.length}"
+            )
+        if self.length > _ManifestLayout.EXPECTED_MAX_LENGTH:
+            raise ValueError(
+                f"Invalid length. Expected at most {_ManifestLayout.EXPECTED_MAX_LENGTH}. Got {self.length}"
+            )
+        if self.signature_size != len(self.signature):
+            raise ValueError(
+                f"Invalid signature size must match len(self.signature). Expected {len(self.signature)}. Got {self.signature_size}"
+            )
+
+    def is_valid(self) -> bool:
+        """Returns true if the manifest appears to be well-formed."""
+        return (
+            self.magic == _ManifestLayout.EXPECTED_MAGIC
+            and self.length >= _ManifestLayout.EXPECTED_MIN_LENGTH
+            and self.length <= _ManifestLayout.EXPECTED_MAX_LENGTH
+            and self.signature_size >= 0
+            and self.signature_size <= _ManifestLayout.MAX_SIGNATURE_SIZE
+            and self.program_offset >= 0
+            and len(self.signature) == self.signature_size
+        )
+
+    def to_bytes(self) -> bytes:
+        """"Returns the binary representation of the Manifest. Written bottom up
+        to allow for BC considerations. The compatibility-preserving way to make
+        changes is to increase the header's length field and add new fields at
+        the top. This means we can always check the last 8 bytes for the magic
+        and size, and then load the full footer."
+        """
+        if not self.is_valid():
+            raise ValueError("Cannot serialize an invalid manifest")
+
+        data: bytes = (
+            # bytes: Signature unique ID for the data the manifest was appended to.
+            self.signature
+            # actual size of the signature
+            + self.signature_size.to_bytes(4, byteorder=_MANIFEST_BYTEORDER)
+            # uint32_t: Any padding required to align the manifest.
+            + self.padding_size.to_bytes(4, byteorder=_MANIFEST_BYTEORDER)
+            # uint64_t: Size in bytes between the manifest and the data it was appended to.
+            + self.program_offset.to_bytes(8, byteorder=_MANIFEST_BYTEORDER)
+            # uint32_t: Actual size of this manifest.
+            + self.length.to_bytes(4, byteorder=_MANIFEST_BYTEORDER)
+            # Manifest magic. This lets consumers detect whether the
+            # manifest was inserted or not. Always use the proper magic value
+            # (i.e., ignore self.magic) since there's no reason to create an
+            # invalid manifest.
+            + self.EXPECTED_MAGIC
+        )
+        return data
+
+    @staticmethod
+    def from_bytes(data: bytes) -> "_ManifestLayout":
```

Comment:
For large files you have to read the whole thing?

Reply:
Could add more methods like from file that just load the last MAX_SIZE bytes. Or stream.

Reply:
I dont really expect people to be verifying the hash in python though. Its mostly just there for testing.